File: packages/expo-updates/android/src/main/java/expo/modules/updates/db/dao/UpdateDao.java
package expo.modules.updates.db.dao;

import androidx.room.Delete;
import androidx.room.Update;
import expo.modules.updates.db.enums.UpdateStatus;
import expo.modules.updates.db.entity.AssetEntity;
import expo.modules.updates.db.entity.UpdateEntity;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.UUID;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Transaction;

@Dao
public abstract class UpdateDao {
  /**
   * for private use only
   * must be marked public for Room
   * so we use the underscore to discourage use
   */
  @Query("SELECT * FROM updates WHERE scope_key = :scopeKey AND status IN (:statuses);")
  public abstract List<UpdateEntity> _loadUpdatesForProjectWithStatuses(String scopeKey, List<UpdateStatus> statuses);

  @Query("SELECT * FROM updates WHERE id = :id;")
  public abstract List<UpdateEntity> _loadUpdatesWithId(UUID id);

  @Query("SELECT assets.* FROM assets INNER JOIN updates ON updates.launch_asset_id = assets.id WHERE updates.id = :id;")
  public abstract AssetEntity _loadLaunchAsset(UUID id);

  @Query("UPDATE updates SET keep = 1 WHERE id = :id;")
  public abstract void _keepUpdate(UUID id);

  @Query("UPDATE updates SET status = :status WHERE id = :id;")
  public abstract void _markUpdateWithStatus(UpdateStatus status, UUID id);

  @Update
  public abstract void _updateUpdate(UpdateEntity update);

  @Query("UPDATE updates SET status = :status WHERE id IN (" +
          "SELECT DISTINCT update_id FROM updates_assets WHERE asset_id IN (:missingAssetIds));")
  public abstract void _markUpdatesWithMissingAssets(List<Long> missingAssetIds, UpdateStatus status);


  /**
   * for public use
   */

  @Query("SELECT * FROM updates;")
  public abstract List<UpdateEntity> loadAllUpdates();

  public List<UpdateEntity> loadLaunchableUpdatesForScope(String scopeKey) {
    return _loadUpdatesForProjectWithStatuses(scopeKey, Arrays.asList(UpdateStatus.READY, UpdateStatus.EMBEDDED, UpdateStatus.DEVELOPMENT));
  }

  @Query("SELECT * FROM updates WHERE status = :status;")
  public abstract List<UpdateEntity> loadAllUpdatesWithStatus(UpdateStatus status);

  public UpdateEntity loadUpdateWithId(UUID id) {
    List<UpdateEntity> updateEntities = _loadUpdatesWithId(id);
    return updateEntities.size() > 0 ? updateEntities.get(0) : null;
  }

  public AssetEntity loadLaunchAsset(UUID id) {
    AssetEntity assetEntity = _loadLaunchAsset(id);
    assetEntity.isLaunchAsset = true;
    return assetEntity;
  }

  @Insert
  public abstract void insertUpdate(UpdateEntity update);

  public void setUpdateScopeKey(UpdateEntity update, String newScopeKey) {
    update.scopeKey = newScopeKey;
    _updateUpdate(update);
  }

  @Transaction
  public void markUpdateFinished(UpdateEntity update, boolean hasSkippedEmbeddedAssets) {
    UpdateStatus statusToMark = UpdateStatus.READY;
    if (update.status == UpdateStatus.DEVELOPMENT) {
      statusToMark = UpdateStatus.DEVELOPMENT;
    } else if (hasSkippedEmbeddedAssets) {
      statusToMark = UpdateStatus.EMBEDDED;
    }
    _markUpdateWithStatus(statusToMark, update.id);
    _keepUpdate(update.id);
  }

  public void markUpdateFinished(UpdateEntity update) {
    markUpdateFinished(update, false);
  }

  public void markUpdateAccessed(UpdateEntity update) {
    update.lastAccessed = new Date();
    _updateUpdate(update);
  }

  public void markUpdatesWithMissingAssets(List<AssetEntity> missingAssets) {
    List<Long> missingAssetIds = new ArrayList<>();
    for (AssetEntity asset : missingAssets) {
      missingAssetIds.add(asset.id);
    }
    _markUpdatesWithMissingAssets(missingAssetIds, UpdateStatus.PENDING);
  }

  @Delete
  public abstract void deleteUpdates(List<UpdateEntity> updates);
}

File: packages/expo-updates/android/src/main/java/expo/modules/updates/db/dao/UpdateDao.kt
package expo.modules.updates.db.dao

import androidx.room.*
import expo.modules.updates.db.entity.AssetEntity
import expo.modules.updates.db.entity.UpdateEntity
import expo.modules.updates.db.enums.UpdateStatus
import java.util.*

/**
 * Utility class for accessing and modifying data in SQLite relating to updates.
 */
@Dao
abstract class UpdateDao {
  // if an update has successfully launched at least once, we treat it as launchable
  // even if it has also failed to launch at least once
  @Query("SELECT * FROM updates WHERE scope_key = :scopeKey AND (successful_launch_count > 0 OR failed_launch_count < 1) AND status IN (:statuses);")
  protected abstract fun loadLaunchableUpdatesForProjectWithStatuses(scopeKey: String, statuses: List<UpdateStatus>): List<UpdateEntity>

  @Query("SELECT * FROM updates WHERE id = :id;")
  protected abstract fun loadUpdatesWithId(id: UUID): List<UpdateEntity>

  @Query("SELECT assets.* FROM assets INNER JOIN updates ON updates.launch_asset_id = assets.id WHERE updates.id = :updateId;")
  protected abstract fun loadLaunchAssetForUpdateInternal(updateId: UUID): AssetEntity?

  @Query("UPDATE updates SET keep = 1 WHERE id = :id;")
  protected abstract fun keepUpdate(id: UUID)

  @Query("UPDATE updates SET status = :status WHERE id = :id;")
  protected abstract fun markUpdateWithStatus(status: UpdateStatus, id: UUID)

  @Query(
    "UPDATE updates SET status = :status WHERE id IN (" +
      "SELECT DISTINCT update_id FROM updates_assets WHERE asset_id IN (:missingAssetIds));"
  )
  protected abstract fun markUpdatesWithMissingAssets(missingAssetIds: List<Long>, status: UpdateStatus)

  /**
   * for public use
   */
  @Query("SELECT * FROM updates;")
  abstract fun loadAllUpdates(): List<UpdateEntity>

  fun loadLaunchableUpdatesForScope(scopeKey: String): List<UpdateEntity> {
    return loadLaunchableUpdatesForProjectWithStatuses(
      scopeKey,
      listOf(UpdateStatus.READY, UpdateStatus.EMBEDDED, UpdateStatus.DEVELOPMENT)
    )
  }

  @Query("SELECT * FROM updates WHERE status = :status;")
  abstract fun loadAllUpdatesWithStatus(status: UpdateStatus): List<UpdateEntity>

  @Query("SELECT id FROM updates WHERE status = :status;")
  abstract fun loadAllUpdateIdsWithStatus(status: UpdateStatus): List<UUID>

  @Query("SELECT id FROM updates WHERE failed_launch_count > 0 ORDER BY commit_time DESC LIMIT 5;")
  abstract fun loadRecentUpdateIdsWithFailedLaunch(): List<UUID>

  fun loadUpdateWithId(id: UUID): UpdateEntity? {
    val updateEntities = loadUpdatesWithId(id)
    return if (updateEntities.isNotEmpty()) updateEntities[0] else null
  }

  fun loadLaunchAssetForUpdate(updateId: UUID): AssetEntity? {
    return loadLaunchAssetForUpdateInternal(updateId)?.apply {
      isLaunchAsset = true
    }
  }

  @Insert
  abstract fun insertUpdate(update: UpdateEntity)

  fun setUpdateScopeKey(update: UpdateEntity, newScopeKey: String) {
    update.scopeKey = newScopeKey
    _setUpdateScopeKeyInternal(update.id, newScopeKey)
  }

  @Query("UPDATE updates SET scope_key = :newScopeKey WHERE id = :id;")
  abstract fun _setUpdateScopeKeyInternal(id: UUID, newScopeKey: String)

  fun setUpdateCommitTime(update: UpdateEntity, commitTime: Date) {
    update.commitTime = commitTime
    setUpdateCommitTimeInternal(update.id, commitTime)
  }

  @Query("UPDATE updates SET commit_time = :commitTime WHERE id = :id;")
  abstract fun setUpdateCommitTimeInternal(id: UUID, commitTime: Date)

  @Transaction
  open fun markUpdateFinished(update: UpdateEntity, hasSkippedEmbeddedAssets: Boolean) {
    var statusToMark = UpdateStatus.READY
    if (update.status === UpdateStatus.DEVELOPMENT) {
      statusToMark = UpdateStatus.DEVELOPMENT
    } else if (hasSkippedEmbeddedAssets) {
      statusToMark = UpdateStatus.EMBEDDED
    }
    markUpdateWithStatus(statusToMark, update.id)
    keepUpdate(update.id)
  }

  fun markUpdateFinished(update: UpdateEntity) {
    markUpdateFinished(update, false)
  }

  fun markUpdateAccessed(update: UpdateEntity) {
    val newLastAccessed = Date()
    update.lastAccessed = newLastAccessed
    markUpdateAccessedInternal(update.id, newLastAccessed)
  }

  @Query("UPDATE updates SET last_accessed = :lastAccessed WHERE id = :id;")
  protected abstract fun markUpdateAccessedInternal(id: UUID, lastAccessed: Date)

  fun incrementSuccessfulLaunchCount(update: UpdateEntity) {
    update.successfulLaunchCount++
    incrementSuccessfulLaunchCountInternal(update.id)
  }

  @Query("UPDATE updates SET successful_launch_count = successful_launch_count + 1 WHERE id = :id;")
  protected abstract fun incrementSuccessfulLaunchCountInternal(id: UUID)

  fun incrementFailedLaunchCount(update: UpdateEntity) {
    update.failedLaunchCount++
    incrementFailedLaunchCountInternal(update.id)
  }

  @Query("UPDATE updates SET failed_launch_count = failed_launch_count + 1 WHERE id = :id;")
  abstract fun incrementFailedLaunchCountInternal(id: UUID)

  fun markUpdatesWithMissingAssets(missingAssets: List<AssetEntity>) {
    val missingAssetIds = mutableListOf<Long>()
    for (asset in missingAssets) {
      missingAssetIds.add(asset.id)
    }
    markUpdatesWithMissingAssets(missingAssetIds, UpdateStatus.PENDING)
  }

  @Delete
  abstract fun deleteUpdates(updates: List<UpdateEntity>)
}
